File: mobile/android/app/src/main/kotlin/com/masim05/bloodpressure/mobile/adapters/api/HttpApiClient.kt
package com.masim05.bloodpressure.mobile.adapters.api

import com.masim05.bloodpressure.mobile.core.model.AppResult
import com.masim05.bloodpressure.mobile.core.model.ArmSide
import com.masim05.bloodpressure.mobile.core.model.HistoryFilter
import com.masim05.bloodpressure.mobile.core.model.Measurement
import com.masim05.bloodpressure.mobile.core.model.MeasurementDetail
import com.masim05.bloodpressure.mobile.core.model.MeasurementImage
import com.masim05.bloodpressure.mobile.core.model.MeasurementStatus
import com.masim05.bloodpressure.mobile.core.model.MobileUser
import com.masim05.bloodpressure.mobile.core.model.Session
import com.masim05.bloodpressure.mobile.core.ports.AuthGateway
import com.masim05.bloodpressure.mobile.core.ports.HistoryGateway
import com.masim05.bloodpressure.mobile.core.ports.MeasurementDetailGateway
import com.masim05.bloodpressure.mobile.core.ports.MeasurementUploadGateway
import java.io.ByteArrayOutputStream
import java.io.File
import java.net.HttpURLConnection
import java.net.URI
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

class HttpApiClient(
    private val baseUrl: String,
    private val fallbackApiMessage: String,
    private val networkMessage: String,
    private val timeoutMessage: String,
    private val parseMessage: String,
) : AuthGateway, HistoryGateway, MeasurementUploadGateway, MeasurementDetailGateway {
    fun fetchMeasurementImage(imageUrl: String, authorization: String): AppResult<ByteArray> = runCatching {
        val response = requestBytes(url = imageUrl, method = "GET", authorization = authorization, accept = "image/*")
        if (response.status in 200..299) {
            AppResult.Success(response.body)
        } else {
            AppResult.Failure(ApiErrorMapper.fromApiBody(response.body.toString(StandardCharsets.UTF_8), fallbackApiMessage))
        }
    }.getOrElse { AppResult.Failure(ApiErrorMapper.fromThrowable(it, networkMessage, timeoutMessage, parseMessage)) }

    override fun signIn(email: String, password: String): AppResult<Session> = authenticate("/api/v1/signin", email, password)

    override fun logIn(email: String, password: String): AppResult<Session> = authenticate("/api/v1/login", email, password)

    override fun upload(session: Session, image: MeasurementImage): AppResult<String> = runCatching {
        val boundary = "----BloodPressureTrackerBoundary"
        val imageBytes = readImageBytes(image)
        val filename = resolveFilename(image.uri)
        val body = ByteArrayOutputStream().apply {
            write("--$boundary\r\n".toByteArray())
            write("Content-Disposition: form-data; name=\"image\"; filename=\"$filename\"\r\n".toByteArray())
            write("Content-Type: ${image.mimeType}\r\n\r\n".toByteArray())
            write(imageBytes)
            write("\r\n--$boundary--\r\n".toByteArray())
        }.toByteArray()
        val response = request(
            path = "/api/v1/measurements",
            method = "POST",
            body = body,
            contentType = "multipart/form-data; boundary=$boundary",
            authorization = session.authorizationHeader,
        )
        if (response.status in 200..299) {
            AppResult.Success(extractJsonString(response.body, "id") ?: "")
        } else {
            AppResult.Failure(ApiErrorMapper.fromApiBody(response.body, fallbackApiMessage))
        }
    }.getOrElse { AppResult.Failure(ApiErrorMapper.fromThrowable(it, networkMessage, timeoutMessage, parseMessage)) }

    override fun list(session: Session, filter: HistoryFilter): AppResult<List<Measurement>> = runCatching {
        val query = buildList {
            add("page=${filter.page}")
            add("pageSize=${filter.pageSize}")
            if (filter.from.isNotBlank()) add("from=${url(filter.from)}T00%3A00%3A00.000Z")
            if (filter.to.isNotBlank()) add("to=${url(filter.to)}T23%3A59%3A59.999Z")
        }.joinToString("&")
        val response = request("/api/v1/measurements?$query", "GET", authorization = session.authorizationHeader)
        if (response.status in 200..299) {
            AppResult.Success(parseMeasurements(response.body))
        } else {
            AppResult.Failure(ApiErrorMapper.fromApiBody(response.body, fallbackApiMessage))
        }
    }.getOrElse { AppResult.Failure(ApiErrorMapper.fromThrowable(it, networkMessage, timeoutMessage, parseMessage)) }

    override fun get(session: Session, measurementId: String): AppResult<MeasurementDetail> = runCatching {
        val response = request("/api/v1/measurements/${urlPath(measurementId)}", "GET", authorization = session.authorizationHeader)
        if (response.status in 200..299) {
            AppResult.Success(parseMeasurementDetail(response.body))
        } else {
            AppResult.Failure(ApiErrorMapper.fromApiBody(response.body, fallbackApiMessage))
        }
    }.getOrElse { AppResult.Failure(ApiErrorMapper.fromThrowable(it, networkMessage, timeoutMessage, parseMessage)) }

    override fun save(session: Session, detail: MeasurementDetail): AppResult<MeasurementDetail> = runCatching {
        val response = request(
            path = "/api/v1/measurements/${urlPath(detail.id)}/save",
            method = "POST",
            body = saveMeasurementBody(detail).toByteArray(StandardCharsets.UTF_8),
            contentType = "application/json",
            authorization = session.authorizationHeader,
        )
        if (response.status in 200..299) {
            val saved = parseMeasurementDetail(response.body)
            AppResult.Success(saved.copy(imageUrl = saved.imageUrl.ifBlank { detail.imageUrl }))
        } else {
            AppResult.Failure(ApiErrorMapper.fromApiBody(response.body, fallbackApiMessage))
        }
    }.getOrElse { AppResult.Failure(ApiErrorMapper.fromThrowable(it, networkMessage, timeoutMessage, parseMessage)) }

    private fun authenticate(path: String, email: String, password: String): AppResult<Session> = runCatching {
        val json = "{\"email\":\"${escape(email)}\",\"password\":\"${escape(password)}\"}"
        val response = request(path, "POST", json.toByteArray(), "application/json")
        if (response.status in 200..299) {
            AppResult.Success(parseSession(response.body))
        } else {
            AppResult.Failure(ApiErrorMapper.fromApiBody(response.body, fallbackApiMessage))
        }
    }.getOrElse { AppResult.Failure(ApiErrorMapper.fromThrowable(it, networkMessage, timeoutMessage, parseMessage)) }

    private fun request(
        path: String,
        method: String,
        body: ByteArray? = null,
        contentType: String? = null,
        authorization: String? = null,
    ): HttpResponse {
        val connection = URI.create(baseUrl + path).toURL().openConnection() as HttpURLConnection
        return try {
            connection.requestMethod = method
            connection.connectTimeout = 5_000
            connection.readTimeout = 10_000
            authorization?.let { connection.setRequestProperty("Authorization", it) }
            contentType?.let { connection.setRequestProperty("Content-Type", it) }
            if (body != null) {
                connection.doOutput = true
                connection.outputStream.use { it.write(body) }
            }
            val status = connection.responseCode
            val stream = if (status in 200..299) connection.inputStream else connection.errorStream
            HttpResponse(status, stream?.bufferedReader()?.use { it.readText() }.orEmpty())
        } finally {
            connection.disconnect()
        }
    }

    private fun requestBytes(
        url: String,
        method: String,
        authorization: String? = null,
        accept: String? = null,
    ): HttpBytesResponse {
        val connection = URI.create(url).toURL().openConnection() as HttpURLConnection
        return try {
            connection.requestMethod = method
            connection.connectTimeout = 5_000
            connection.readTimeout = 10_000
            authorization?.let { connection.setRequestProperty("Authorization", it) }
            accept?.let { connection.setRequestProperty("Accept", it) }
            val status = connection.responseCode
            val stream = if (status in 200..299) connection.inputStream else connection.errorStream
            HttpBytesResponse(status, stream?.use { it.readBytes() } ?: ByteArray(0))
        } finally {
            connection.disconnect()
        }
    }

    private fun parseSession(body: String): Session = Session(
        accessToken = requireNotNull(extractJsonString(body, "accessToken")),
        tokenType = extractJsonString(body, "tokenType") ?: "Bearer",
        expiresAt = requireNotNull(extractJsonString(body, "expiresAt")),
        user = MobileUser(
            id = requireNotNull(extractJsonString(body, "id")),
            email = requireNotNull(extractJsonString(body, "email")),
        ),
    )

    private fun parseMeasurements(body: String): List<Measurement> {
        val items = Regex("\\{[^{}]*\\\"id\\\"[^{}]*\\}").findAll(body).map { it.value }.toList()
        return items.mapNotNull { item ->
            val status = extractJsonString(item, "status") ?: return@mapNotNull null
            Measurement(
                id = extractJsonString(item, "id") ?: return@mapNotNull null,
                status = parseStatus(status),
                systolic = extractJsonInt(item, "systolic") ?: 0,
                diastolic = extractJsonInt(item, "diastolic") ?: 0,
                pulse = extractJsonInt(item, "pulse") ?: 0,
                armSide = when (extractJsonString(item, "armSide")) {
                    "left" -> ArmSide.Left
                    "right" -> ArmSide.Right
                    else -> ArmSide.Unknown
                },
                measurementTime = extractJsonString(item, "measurementTime") ?: "",
                savedAt = extractJsonString(item, "savedAt") ?: "",
            )
        }
    }

    private fun parseMeasurementDetail(body: String): MeasurementDetail = MeasurementDetail(
        id = requireNotNull(extractJsonString(body, "id")),
        status = parseStatus(extractJsonString(body, "status")),
        systolic = extractJsonInt(body, "systolic"),
        diastolic = extractJsonInt(body, "diastolic"),
        pulse = extractJsonInt(body, "pulse"),
        armSide = parseArmSide(extractJsonString(body, "armSide")),
        measurementTime = extractJsonString(body, "measurementTime") ?: "",
        savedAt = extractJsonString(body, "savedAt"),
        imageUrl = extractJsonString(body, "imageUrl") ?: "",
        recognitionError = extractJsonString(body, "recognitionError"),
    )

    private fun parseStatus(status: String?): MeasurementStatus = when (status) {
        "recognizing" -> MeasurementStatus.Recognizing
        "recognized" -> MeasurementStatus.Recognized
        "saved" -> MeasurementStatus.Saved
        "failed" -> MeasurementStatus.Failed
        else -> MeasurementStatus.Pending
    }

    private fun parseArmSide(value: String?): ArmSide = when (value) {
        "left" -> ArmSide.Left
        "right" -> ArmSide.Right
        else -> ArmSide.Unknown
    }

    private fun extractJsonString(body: String, field: String): String? =
        Regex("\\\"$field\\\"\\s*:\\s*\\\"([^\\\"]*)\\\"").find(body)?.groupValues?.get(1)

    private fun extractJsonInt(body: String, field: String): Int? =
        Regex("\\\"$field\\\"\\s*:\\s*(\\d+)").find(body)?.groupValues?.get(1)?.toInt()

    private fun escape(value: String): String = value.replace("\\", "\\\\").replace("\"", "\\\"")
    private fun url(value: String): String = URLEncoder.encode(value, StandardCharsets.UTF_8.name())
    private fun urlPath(value: String): String = URLEncoder.encode(value, StandardCharsets.UTF_8.name()).replace("+", "%20")
    private fun saveMeasurementBody(detail: MeasurementDetail): String = buildList {
        detail.systolic?.let { add("\"systolic\":$it") }
        detail.diastolic?.let { add("\"diastolic\":$it") }
        detail.pulse?.let { add("\"pulse\":$it") }
    }.joinToString(separator = ",", prefix = "{", postfix = "}")

    private fun readImageBytes(image: MeasurementImage): ByteArray {
        val imageUri = URI.create(image.uri)
        if (imageUri.scheme == null || imageUri.scheme == "file") {
            val file = if (imageUri.scheme == null) {
                File(image.uri)
            } else {
                File(imageUri)
            }
            require(file.exists() && file.isFile) { "Image file is unavailable" }
            return file.readBytes()
        }
        throw IllegalArgumentException("Unsupported image URI scheme: ${imageUri.scheme}")
    }

    private fun resolveFilename(uri: String): String {
        val parsed = URI.create(uri)
        val fromPath = parsed.path?.substringAfterLast('/')?.takeIf { it.isNotBlank() }
        if (fromPath != null) {
            return fromPath
        }
        return "measurement"
    }

    private data class HttpResponse(val status: Int, val body: String)
    private data class HttpBytesResponse(val status: Int, val body: ByteArray)
}

File: mobile/ios/Blood pressure tracker/Blood pressure tracker/HttpApiClient.swift
// HttpApiClient.swift
// Blood pressure tracker
// HTTP API client mirroring Android's HttpApiClient.kt using URLSession with synchronous dispatch.

import Foundation

class HttpApiClient: AuthGateway, HistoryGateway, MeasurementUploadGateway, MeasurementDetailGateway {
    let baseUrl: String
    private let fallbackApiMessage: String
    private let networkMessage: String
    private let timeoutMessage: String
    private let parseMessage: String

    init(
        baseUrl: String,
        fallbackApiMessage: String = "Unexpected API error",
        networkMessage: String = "Network error",
        timeoutMessage: String = "Timeout",
        parseMessage: String = "Parse error"
    ) {
        self.baseUrl = baseUrl
        self.fallbackApiMessage = fallbackApiMessage
        self.networkMessage = networkMessage
        self.timeoutMessage = timeoutMessage
        self.parseMessage = parseMessage
    }

    // MARK: - AuthGateway

    func signIn(email: String, password: String) -> AppResult<Session> {
        authenticate(path: "/api/v1/signin", email: email, password: password)
    }

    func logIn(email: String, password: String) -> AppResult<Session> {
        authenticate(path: "/api/v1/login", email: email, password: password)
    }

    // MARK: - MeasurementUploadGateway

    func upload(session: Session, image: MeasurementImage) -> AppResult<String> {
        do {
            let boundary = "----BloodPressureTrackerBoundary"
            let imageData = try readImageData(from: image)
            let filename = resolveFilename(from: image.uri)

            var body = Data()
            body += "--\(boundary)\r\n".utf8Data
            body += "Content-Disposition: form-data; name=\"image\"; filename=\"\(filename)\"\r\n".utf8Data
            body += "Content-Type: \(image.mimeType)\r\n\r\n".utf8Data
            body += imageData
            body += "\r\n--\(boundary)--\r\n".utf8Data

            let response = try request(
                path: "/api/v1/measurements",
                method: "POST",
                body: body,
                contentType: "multipart/form-data; boundary=\(boundary)",
                authorization: session.authorizationHeader
            )
            if (200 ... 299).contains(response.statusCode) {
                return .success(extractJsonString(from: response.body, field: "id") ?? "")
            }
            return .failure(ApiErrorMapper.fromApiBody(response.body, fallback: fallbackApiMessage))
        } catch {
            return .failure(ApiErrorMapper.fromError(error, networkMessage: networkMessage, timeoutMessage: timeoutMessage, parseMessage: parseMessage))
        }
    }

    // MARK: - HistoryGateway

    func list(session: Session, filter: HistoryFilter) -> AppResult<[Measurement]> {
        do {
            var queryParts = ["page=\(filter.page)", "pageSize=\(filter.pageSize)"]
            if !filter.from.isEmpty {
                queryParts.append("from=\(urlEncode(filter.from))T00%3A00%3A00.000Z")
            }
            if !filter.to.isEmpty {
                queryParts.append("to=\(urlEncode(filter.to))T23%3A59%3A59.999Z")
            }
            let query = queryParts.joined(separator: "&")
            let response = try request(
                path: "/api/v1/measurements?\(query)",
                method: "GET",
                authorization: session.authorizationHeader
            )
            if (200 ... 299).contains(response.statusCode) {
                return .success(parseMeasurements(response.body))
            }
            return .failure(ApiErrorMapper.fromApiBody(response.body, fallback: fallbackApiMessage))
        } catch {
            return .failure(ApiErrorMapper.fromError(error, networkMessage: networkMessage, timeoutMessage: timeoutMessage, parseMessage: parseMessage))
        }
    }

    // MARK: - MeasurementDetailGateway

    func get(session: Session, measurementId: String) -> AppResult<MeasurementDetail> {
        do {
            let response = try request(
                path: "/api/v1/measurements/\(urlPathEncode(measurementId))",
                method: "GET",
                authorization: session.authorizationHeader
            )
            if (200 ... 299).contains(response.statusCode) {
                guard let detail = parseMeasurementDetail(response.body) else {
                    return .failure(ApiError(code: nil, message: parseMessage, source: .parse))
                }
                return .success(detail)
            }
            return .failure(ApiErrorMapper.fromApiBody(response.body, fallback: fallbackApiMessage))
        } catch {
            return .failure(ApiErrorMapper.fromError(error, networkMessage: networkMessage, timeoutMessage: timeoutMessage, parseMessage: parseMessage))
        }
    }

    func save(session: Session, detail: MeasurementDetail) -> AppResult<MeasurementDetail> {
        do {
            let bodyString = saveMeasurementBody(detail)
            let response = try request(
                path: "/api/v1/measurements/\(urlPathEncode(detail.id))/save",
                method: "POST",
                body: bodyString.utf8Data,
                contentType: "application/json",
                authorization: session.authorizationHeader
            )
            if (200 ... 299).contains(response.statusCode) {
                guard var saved = parseMeasurementDetail(response.body) else {
                    return .failure(ApiError(code: nil, message: parseMessage, source: .parse))
                }
                if saved.imageUrl.isEmpty {
                    saved = saved.withImageUrl(detail.imageUrl)
                }
                return .success(saved)
            }
            return .failure(ApiErrorMapper.fromApiBody(response.body, fallback: fallbackApiMessage))
        } catch {
            return .failure(ApiErrorMapper.fromError(error, networkMessage: networkMessage, timeoutMessage: timeoutMessage, parseMessage: parseMessage))
        }
    }

    // MARK: - Image fetching

    func fetchMeasurementImage(imageUrl: String, authorization: String) -> AppResult<Data> {
        do {
            let response = try requestBytes(
                urlString: imageUrl,
                method: "GET",
                authorization: authorization,
                accept: "image/*"
            )
            if (200 ... 299).contains(response.statusCode) {
                return .success(response.body)
            }
            let errorBody = String(data: response.body, encoding: .utf8) ?? ""
            return .failure(ApiErrorMapper.fromApiBody(errorBody, fallback: fallbackApiMessage))
        } catch {
            return .failure(ApiErrorMapper.fromError(error, networkMessage: networkMessage, timeoutMessage: timeoutMessage, parseMessage: parseMessage))
        }
    }

    // MARK: - Private HTTP primitives

    private struct HttpResponse {
        let statusCode: Int
        let body: String
    }

    private struct HttpBytesResponse {
        let statusCode: Int
        let body: Data
    }

    private func request(
        path: String,
        method: String,
        body: Data? = nil,
        contentType: String? = nil,
        authorization: String? = nil
    ) throws -> HttpResponse {
        let urlString = baseUrl + path
        guard let url = URL(string: urlString) else { throw URLError(.badURL) }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.timeoutInterval = 15
        if let auth = authorization { req.setValue(auth, forHTTPHeaderField: "Authorization") }
        if let ct = contentType { req.setValue(ct, forHTTPHeaderField: "Content-Type") }
        if let b = body { req.httpBody = b }

        var resultBody: String = ""
        var resultStatus: Int = 0
        var resultError: Error?
        let semaphore = DispatchSemaphore(value: 0)

        URLSession.shared.dataTask(with: req) { data, response, error in
            if let e = error {
                resultError = e
            } else if let http = response as? HTTPURLResponse {
                resultStatus = http.statusCode
                resultBody = data.flatMap { String(data: $0, encoding: .utf8) } ?? ""
            }
            semaphore.signal()
        }.resume()
        semaphore.wait()

        if let e = resultError { throw e }
        return HttpResponse(statusCode: resultStatus, body: resultBody)
    }

    private func requestBytes(
        urlString: String,
        method: String,
        authorization: String? = nil,
        accept: String? = nil
    ) throws -> HttpBytesResponse {
        guard let url = URL(string: urlString) else { throw URLError(.badURL) }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.timeoutInterval = 15
        if let auth = authorization { req.setValue(auth, forHTTPHeaderField: "Authorization") }
        if let acc = accept { req.setValue(acc, forHTTPHeaderField: "Accept") }

        var resultData: Data = Data()
        var resultStatus: Int = 0
        var resultError: Error?
        let semaphore = DispatchSemaphore(value: 0)

        URLSession.shared.dataTask(with: req) { data, response, error in
            if let e = error {
                resultError = e
            } else if let http = response as? HTTPURLResponse {
                resultStatus = http.statusCode
                resultData = data ?? Data()
            }
            semaphore.signal()
        }.resume()
        semaphore.wait()

        if let e = resultError { throw e }
        return HttpBytesResponse(statusCode: resultStatus, body: resultData)
    }

    private func authenticate(path: String, email: String, password: String) -> AppResult<Session> {
        do {
            let json = "{\"email\":\"\(escapeJson(email))\",\"password\":\"\(escapeJson(password))\"}"
            let response = try request(
                path: path,
                method: "POST",
                body: json.utf8Data,
                contentType: "application/json"
            )
            if (200 ... 299).contains(response.statusCode) {
                guard let session = parseSession(response.body) else {
                    return .failure(ApiError(code: nil, message: parseMessage, source: .parse))
                }
                return .success(session)
            }
            return .failure(ApiErrorMapper.fromApiBody(response.body, fallback: fallbackApiMessage))
        } catch {
            return .failure(ApiErrorMapper.fromError(error, networkMessage: networkMessage, timeoutMessage: timeoutMessage, parseMessage: parseMessage))
        }
    }

    // MARK: - JSON parsing

    private func parseSession(_ body: String) -> Session? {
        guard let accessToken = extractJsonString(from: body, field: "accessToken"),
              let expiresAt = extractJsonString(from: body, field: "expiresAt"),
              let id = extractJsonString(from: body, field: "id"),
              let email = extractJsonString(from: body, field: "email") else { return nil }
        let tokenType = extractJsonString(from: body, field: "tokenType") ?? "Bearer"
        return Session(
            accessToken: accessToken,
            tokenType: tokenType,
            expiresAt: expiresAt,
            user: MobileUser(id: id, email: email)
        )
    }

    private func parseMeasurements(_ body: String) -> [Measurement] {
        guard let pattern = try? NSRegularExpression(pattern: #"\{[^{}]*"id"[^{}]*\}"#) else { return [] }
        let range = NSRange(body.startIndex..., in: body)
        return pattern.matches(in: body, range: range).compactMap { match -> Measurement? in
            guard let r = Range(match.range, in: body) else { return nil }
            let item = String(body[r])
            guard let id = extractJsonString(from: item, field: "id"),
                  let statusStr = extractJsonString(from: item, field: "status") else { return nil }
            return Measurement(
                id: id,
                status: parseStatus(statusStr),
                systolic: extractJsonInt(from: item, field: "systolic") ?? 0,
                diastolic: extractJsonInt(from: item, field: "diastolic") ?? 0,
                pulse: extractJsonInt(from: item, field: "pulse") ?? 0,
                armSide: parseArmSide(extractJsonString(from: item, field: "armSide")),
                measurementTime: extractJsonString(from: item, field: "measurementTime") ?? "",
                savedAt: extractJsonString(from: item, field: "savedAt") ?? ""
            )
        }
    }

    private func parseMeasurementDetail(_ body: String) -> MeasurementDetail? {
        guard let id = extractJsonString(from: body, field: "id") else { return nil }
        return MeasurementDetail(
            id: id,
            status: parseStatus(extractJsonString(from: body, field: "status")),
            systolic: extractJsonInt(from: body, field: "systolic"),
            diastolic: extractJsonInt(from: body, field: "diastolic"),
            pulse: extractJsonInt(from: body, field: "pulse"),
            armSide: parseArmSide(extractJsonString(from: body, field: "armSide")),
            measurementTime: extractJsonString(from: body, field: "measurementTime") ?? "",
            savedAt: extractJsonString(from: body, field: "savedAt"),
            imageUrl: extractJsonString(from: body, field: "imageUrl") ?? "",
            recognitionError: extractJsonString(from: body, field: "recognitionError")
        )
    }

    private func parseStatus(_ status: String?) -> MeasurementStatus {
        switch status {
        case "recognizing": return .recognizing
        case "recognized": return .recognized
        case "saved": return .saved
        case "failed": return .failed
        default: return .pending
        }
    }

    private func parseArmSide(_ value: String?) -> ArmSide {
        switch value {
        case "left": return .left
        case "right": return .right
        default: return .unknown
        }
    }

    func extractJsonString(from body: String, field: String) -> String? {
        let escapedField = NSRegularExpression.escapedPattern(for: field)
        guard let pattern = try? NSRegularExpression(
            pattern: "\"" + escapedField + "\"\\s*:\\s*\"([^\"]*)\""
        ) else { return nil }
        let range = NSRange(body.startIndex..., in: body)
        guard let match = pattern.firstMatch(in: body, range: range),
              let valueRange = Range(match.range(at: 1), in: body) else { return nil }
        return String(body[valueRange])
    }

    func extractJsonInt(from body: String, field: String) -> Int? {
        let escapedField = NSRegularExpression.escapedPattern(for: field)
        guard let pattern = try? NSRegularExpression(
            pattern: "\"" + escapedField + "\"\\s*:\\s*(\\d+)"
        ) else { return nil }
        let range = NSRange(body.startIndex..., in: body)
        guard let match = pattern.firstMatch(in: body, range: range),
              let valueRange = Range(match.range(at: 1), in: body) else { return nil }
        return Int(String(body[valueRange]))
    }

    private func escapeJson(_ value: String) -> String {
        value.replacingOccurrences(of: "\\", with: "\\\\")
             .replacingOccurrences(of: "\"", with: "\\\"")
    }

    private func urlEncode(_ value: String) -> String {
        value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? value
    }

    private func urlPathEncode(_ value: String) -> String {
        (value.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? value)
            .replacingOccurrences(of: "+", with: "%20")
    }

    private func saveMeasurementBody(_ detail: MeasurementDetail) -> String {
        var parts: [String] = []
        if let s = detail.systolic { parts.append("\"systolic\":\(s)") }
        if let d = detail.diastolic { parts.append("\"diastolic\":\(d)") }
        if let p = detail.pulse { parts.append("\"pulse\":\(p)") }
        parts.append("\"armSide\":\"\(serializeArmSide(detail.armSide))\"")
        return "{\(parts.joined(separator: ","))}"
    }

    private func serializeArmSide(_ armSide: ArmSide) -> String {
        switch armSide {
        case .left: return "left"
        case .right: return "right"
        case .unknown: return "unknown"
        }
    }

    private func readImageData(from image: MeasurementImage) throws -> Data {
        guard let url = URL(string: image.uri) else {
            throw NSError(domain: NSCocoaErrorDomain, code: NSFileNoSuchFileError)
        }
        let fileUrl: URL
        if url.scheme == nil || url.scheme == "file" {
            fileUrl = url.scheme == nil ? URL(fileURLWithPath: image.uri) : url
        } else {
            throw NSError(domain: NSCocoaErrorDomain, code: NSFileNoSuchFileError)
        }
        return try Data(contentsOf: fileUrl)
    }

    private func resolveFilename(from uri: String) -> String {
        URL(string: uri)?.lastPathComponent.isEmpty == false
            ? URL(string: uri)!.lastPathComponent
            : "measurement"
    }
}

// MARK: - Helpers

private extension String {
    var utf8Data: Data { Data(utf8) }
}

private func += (lhs: inout Data, rhs: Data) { lhs.append(rhs) }
